File: src/dataset/cpp/leap_Exercism_johnngugi/gemini/GeminiFlash_leap_v2.cpp
#include "leap.h"

bool leap::is_leap_year(int year)
{
    // According to the Gregorian calendar rules:
    // A year is a leap year if it is divisible by 4,
    // unless it is divisible by 100 but not by 400.

    // 1. If the year is not divisible by 4, it's definitely not a leap year.
    // Using a bitwise AND operation for divisibility by 4 (a power of 2)
    // is typically more efficient than the modulo operator.
    if ((year & 3) != 0) { // Equivalent to (year % 4 != 0)
        return false;
    }

    // 2. If the year is divisible by 4, now apply the century rules.
    // If it's divisible by 100...
    if ((year % 100) == 0) {
        // ...then it's a leap year only if it's also divisible by 400.
        return (year % 400) == 0;
    }

    // 3. If the year is divisible by 4 but not by 100, it is a leap year.
    return true;
}

File: src/dataset/go/leap_Exercism_go_thinkverse/gemini/GeminiFlash_leap_v2.go
// Package leap contains a function for checking if a year is a leap year.
package leap

// IsLeapYear checks if a year is a leap year.
//
// Optimized for reduced runtime energy consumption by:
// - Replacing modulo 4 with a bitwise AND operation, which is typically faster.
// - Maintaining an early exit for common non-leap year cases.
func IsLeapYear(year int) bool {
	// A year must be divisible by 4.
	// Using bitwise AND for divisibility by a power of 2 (like 4) is generally
	// more efficient than the modulo operator. (year & 3) is equivalent to year % 4.
	if (year & 3) != 0 { // If not divisible by 4
		return false
	}

	// For years divisible by 4, apply the century rule:
	// If it's divisible by 100, it must also be divisible by 400.
	// This means, if (year % 100 == 0 AND year % 400 != 0), it's NOT a leap year.
	if year%100 == 0 && year%400 != 0 {
		return false
	}

	// If it passed both checks, it's a leap year.
	return true
}